src/ctap2/commands/bio_enrollment.rs

```diff
@@ -0,0 +1,540 @@
+        }
+
+        map.end()
+    }
+}
+
+impl PinUvAuthCommand for BioEnrollment {
+    fn pin(&self) -> &Option<Pin> {
+        &self.pin
+    }
+
+    fn set_pin(&mut self, pin: Option<Pin>) {
+        self.pin = pin;
+    }
+
+    fn get_rp_id(&self) -> Option<&String> {
+        None
+    }
+
+    fn set_pin_uv_auth_param(
+        &mut self,
+        pin_uv_auth_token: Option<PinUvAuthToken>,
+    ) -> Result<(), AuthenticatorError> {
+        let mut param = None;
+        if let Some(token) = pin_uv_auth_token {
+            // pinUvAuthParam (0x04): the result of calling
+            // authenticate(pinUvAuthToken, fingerprint (0x01) || uint8(subCommand) || subCommandParams).
+            let (id, params) = self.subcommand.to_id_and_param();
+            let modality = self.modality as u8;
+            let mut data = vec![modality, id];
+            if params.has_some() {
+                data.extend(to_vec(&params).map_err(CommandError::Serializing)?);
+            }
+            param = Some(token.clone().derive(&data).map_err(CommandError::Crypto)?);
+            self.pin_uv_auth_token = Some(token);
+        }
+        self.pin_uv_auth_param = param;
+        Ok(())
+    }
+
+    fn can_skip_user_verification(
+        &mut self,
+        _info: &crate::AuthenticatorInfo,
+        _uv: UserVerificationRequirement,
+    ) -> bool {
+        // "discouraged" does not exist for BioEnrollment
+        false
+    }
+
+    fn set_uv_option(&mut self, _uv: Option<bool>) {
+        /* No-op */
+    }
+
+    fn get_pin_uv_auth_param(&self) -> Option<&PinUvAuthParam> {
+        self.pin_uv_auth_param.as_ref()
+    }
+}
+
+impl RequestCtap2 for BioEnrollment {
+    type Output = BioEnrollmentResponse;
+
+    fn command(&self) -> Command {
+        if self.use_legacy_preview {
+            Command::BioEnrollmentPreview
+        } else {
+            Command::BioEnrollment
+        }
+    }
+
+    fn wire_format(&self) -> Result<Vec<u8>, HIDError> {
+        let output = to_vec(&self).map_err(CommandError::Serializing)?;
+        trace!("client subcommmand: {:04X?}", &output);
+        Ok(output)
+    }
+
+    fn handle_response_ctap2<Dev>(
+        &self,
+        _dev: &mut Dev,
+        input: &[u8],
+    ) -> Result<Self::Output, HIDError>
+    where
+        Dev: FidoDevice,
+    {
+        if input.is_empty() {
+            return Err(CommandError::InputTooSmall.into());
+        }
+
+        let status: StatusCode = input[0].into();
+        if status.is_ok() {
+            if input.len() > 1 {
+                trace!("parsing bio enrollment response data: {:#04X?}", &input);
+                let bio_enrollment =
+                    from_slice(&input[1..]).map_err(CommandError::Deserializing)?;
+                Ok(bio_enrollment)
+            } else {
+                // Some subcommands return only an OK-status without any data
+                Ok(BioEnrollmentResponse::default())
```

Comment:
We can probably annotate things such that serde will deserialize the default value from the empty slice. I'm not sure of the exact syntax.

Reply:
I didn't find a nice way to do this.